invenio_previewer/previewerext/csv_dthreejs.py

```diff
@@ -1,79 +1,83 @@
+# -*- coding: utf-8 -*-
+#
 # This file is part of Invenio.
-# Copyright (C) 2013, 2014, 2014, 2015 CERN.
+# Copyright (C) 2016 CERN.
 #
-# Invenio is free software; you can redistribute it and/or
-# modify it under the terms of the GNU General Public License as
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
 # published by the Free Software Foundation; either version 2 of the
 # License, or (at your option) any later version.
 #
-# Invenio is distributed in the hope that it will be useful, but
-# WITHOUT ANY WARRANTY; without even the implied warranty of
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 # General Public License for more details.
 #
 # You should have received a copy of the GNU General Public License
-# along with Invenio; if not, write to the Free Software Foundation, Inc.,
-# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
 
 """Render Bar chart from csv file."""
 
 import csv
+import os
 
 from chardet.universaldetector import UniversalDetector
 
-from flask import current_app, render_template, request
+from flask import current_app, render_template
 
-from invenio_ext.cache import cache
 
-
-@cache.memoize(timeout=172800)
-def validate_csv(f):
+def validate_csv(document):
     """Return dialect information about given csv file."""
-    with open(f.fullpath, 'rU') as csvfile:
+    with open(document.document.uri, 'rU') as csvfile:
         is_valid = False
         try:
             dialect = csv.Sniffer().sniff(csvfile.read(1024))
         except Exception as e:
             current_app.logger.debug(
-                'File %s is not valid CSV: %s' % (f.name + f.superformat, e))
+                'File %s is not valid CSV: %s' % (document.get_filename(), e))
             return {
                 'delimiter': '',
                 'encoding': '',
                 'is_valid': is_valid
             }
-        u = UniversalDetector()
+        universal_detector = UniversalDetector()
         dialect.strict = True
         csvfile.seek(0)
         reader = csv.reader(csvfile, dialect)
         try:
             for row in reader:
-                u.feed(dialect.delimiter.join(row))
+                universal_detector.feed(
+                    dialect.delimiter.join(row).encode('utf-8'))
             is_valid = True
         except csv.Error as e:
             current_app.logger.debug(
```

Comment:
I was not able to create a test throwing this exception. The file is read line by line. Each line is split by the delimiter and it is not fed if the length of the result array does not match with the length of the first line. If you are using a "strange" delimiter, the code breaks in a previous line. It is the only line not tested and IMHO it should be removed.